File: javascript/count-the-number-of-fair-pairs.js
const binarySearch = (array, predicate, from = 0, to = array.length) => {
    let lo = from;
    let hi = to;
    
    while (lo < hi) {
        let mi = Math.floor((lo + hi) / 2);
        if (predicate(array[mi])) {
            hi = mi;
        } else {
            lo = mi + 1;
        }
    }
    return lo;
};

/**
 * @param {number[]} nums
 * @param {number} lower
 * @param {number} upper
 * @return {number}
 */
const countFairPairs = (nums, lower, upper) => {
    nums.sort((a, b) => a - b);
    let pairs = 0;
    for (let i = 0; i < nums.length; i++) {
        let to = binarySearch(nums, (otherNum) => nums[i] + otherNum > upper);
        let from = binarySearch(nums, (otherNum) => nums[i] + otherNum >= lower);
        pairs += to - from;
        if (from <= i && i <= to) {
            pairs -= 1;
        }
    }
    return Math.ceil(pairs / 2);
};

File: python3/count-the-number-of-fair-pairs.py
from typing import Callable

class Solution:
    def countFairPairs(self, nums: List[int], lower: int, upper: int) -> int:
        nums.sort()
        fair_pairs = 0
        for i, n in enumerate(nums):
            first = self.binary_search(i + 1, len(nums), lambda j: n + nums[j] >= lower)
            last = self.binary_search(first, len(nums), lambda j: n + nums[j] > upper)
            fair_pairs += last - first
        return fair_pairs

    def binary_search(self, start: int, end: int, predicate: Callable[[int], bool]) -> int:
        lo = start
        hi = end
        while lo < hi:
            mi = (lo + hi) // 2
            if predicate(mi):
                hi = mi
            else:
                lo = mi + 1
        return lo
